### PR TITLE
[EDGORDERS-27] - Honor Accept header

### PR DESCRIPTION
**Pull request by AndreiMPV**
## Purpose
[EDGORDERS-27](https://issues.folio.org/browse/EDGORDERS-27)
Return response according Accept header 

## Approach
Main idea that edge-order become a proxy.
* RAML changed and xml and json now are valid response format
* Rewrite logic
* Update tests

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -38,6 +38,8 @@
 
     <!-- the main class -->
     <exec.mainClass>org.folio.edge.orders.MainVerticle</exec.mainClass>
+    <sonar.exclusions>**/models/**.java</sonar.exclusions>
+    <sonar.test.exclusions>**/*Test.java</sonar.test.exclusions>
```
Comment (craigmcnally) [inline pom.xml:42]: Unless there's a really good reason for this we shouldn't be doing it.

In general you want to limit the scope of exclusions a much as possible, e.g. to a given line for a specific rule.  Even then it's good habit to add a comment explaining why it's being skipped/excluded.

File: ramls/edge-orders.raml [hunk 2]
```diff
@@ -10,33 +10,46 @@ types:
 /orders/validate:
   displayName: Validate
   get:
-    description: Validate that the API Key provided can be used to place an order
+    description: Validate that the API Key provided can be used to place an order. The response body depends on the Vendor (GOBI or other)
```
Comment (craigmcnally) [inline ramls/edge-orders.raml:13]: it would be better to say that it depends on the type query argument in the request

File: ramls/edge-orders.raml [hunk 3]
```diff
@@ -45,70 +58,99 @@ types:
...
+        type: string
 /orders:
   displayName: Place Order
   post:
-    description: Place an order
+    description: Place an order. The response body depends on the Vendor (GOBI or other)
+    headers:
+      Accept:
+        displayName: Accept
+        description: either application/xml or application/json
```
Comment (craigmcnally) [inline ramls/edge-orders.raml:115]: Why not just leave it open instead of putting these limitations in place?  What happens if a module comes along that needs to support application/octet-stream?

File: ramls/edge-orders.raml [hunk 4]
```diff
@@ -45,70 +58,99 @@ types:
...
   post:
-    description: Validate that the API Key provided can be used to place an order
+    description: Validate that the API Key provided can be used to place an order. The response body depends on the Vendor (GOBI or other)
     responses:
        200:
         description: Success
         body:
-          application/xml:
-           example: "<test>POST - OK</test>"
+         application/xml:
```
Comment (craigmcnally) [inline ramls/edge-orders.raml:66]: Just like responses from the order placement endpoint, these should not be prescribed by edge-orders...

File: ramls/edge-orders.raml [hunk 5]
```diff
@@ -45,70 +58,99 @@ types:
...
+          example: |
+            {
+              "test" : "POST - OK"
+            }
        400:
         description: Bad Request
         body:
           application/xml:
-            type: orders
+            example: !include examples/error.xml
```
Comment (craigmcnally) [inline ramls/edge-orders.raml:77]: Following my earlier comments...  This is the one place where it gets a little tricky since the error response might be generated by edge-orders itself instead of being a proxied response.  In this case I think what you have here is fine.  Maybe also include a text/plain as a fallback in the case where neither xml or json are acceptable.

File: src/test/java/org/folio/edge/orders/MainVerticleTest.java [hunk 6]
```diff
@@ -458,4 +481,34 @@ public void testPlaceOrderTimeout(TestContext context) throws MalformedApiKeyExc
...
+      .then()
+      .statusCode(201)
+      .extract()
+      .response();
+
+    assertEquals(StringUtils.EMPTY, resp.body().asString());
+  }
+
+  @Test
+  public void testShouldCatchExceptionIfVendorResponsInvalide(TestContext context) throws JsonProcessingException {
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/orders/MainVerticleTest.java:502]: typo in test name "Invalide"

Comment (craigmcnally) [reply src/test/java/org/folio/edge/orders/MainVerticleTest.java:502]: Also, does this really throw JsonProcessingException?

File: pom.xml (hunk 1 above)
Comment (AndreiMPV) [reply pom.xml:42]: If don't skip it, then Sonar scan our test classes. As result our coverage will fall and development speed will slow down significantly. 
No make sense to check test classes for quality. And  also early test classes were skipped, but some time ago started to include in Sonar check.

File: ramls/edge-orders.raml (hunk 3 above)
Comment (AndreiMPV) [reply ramls/edge-orders.raml]: Replace with free header fron vendor

File: ramls/edge-orders.raml (hunk 2 above)
Comment (AndreiMPV) [reply ramls/edge-orders.raml]: Thanks. Changed

File: ramls/edge-orders.raml (hunk 5 above)
Comment (AndreiMPV) [reply ramls/edge-orders.raml:66]: text/plain added in all error responses

File: src/test/java/org/folio/edge/orders/MainVerticleTest.java (hunk 6 above)
Comment (AndreiMPV) [reply src/test/java/org/folio/edge/orders/MainVerticleTest.java]: Thanks You are right. Tests was changed, but forgot change a name. Fixed

File: src/main/java/org/folio/edge/orders/OrdersHandler.java [hunk 7]
```diff
@@ -99,107 +98,98 @@ protected void handleResponse(RoutingContext ctx, HttpClientResponse resp) {
...
-
-            if (logger.isDebugEnabled()) {
-              logger.debug("status: {} response: {}", + resp.statusCode(), respBody);
-            }
-
             try {
-              String xml = StringUtils.EMPTY;
-              if (respBody.matches(VALIDATE_SUCCESS)) {
-                xml = respBody;
+              if (respBody.matches(VALIDATE_GOBI_XML_SUCCESS)) {
```
Comment (craigmcnally) [inline src/main/java/org/folio/edge/orders/OrdersHandler.java:106]: Why do we need to do any of this?  Can't we just send back the response as-is?  The point of this feature is to avoid having to make code changes when adding new vendor integrations (i.e. types).  If we continue to perform response validation and/or parse the response that becomes difficult.  It's probably better to just trust that the vendor-specific backend module is returning a valid/well-formed response and let the edge API be "dumb"... just pass the response along.

Comment (craigmcnally) [reply src/main/java/org/folio/edge/orders/OrdersHandler.java:106]: Also note that in the case of error responses, I don't think we want to try to map things to a common error schema, regardless of the content type....  Different vendor systems will undoubtedly have differing expectations about what an error response looks like.  So, just like the non-error responses, I think we want to keep this generic and just pass the response from the backend to the client as-is.  

The one exception is when something fails in the edge module itself e.g. invalid API key, or inability to get credentials/logic, etc.  In these cases I think we probably need to allow for some error templates to be supplied via configuration and do some substitution, e.g. an XML snippet can be provided for "invalid API key" responses.  The snippet could have placeholders for things like {message}, {code}, etc.   This however sounds like another story.

**In the short term, I would  just continue to use the error responses you have defined here for these types of errors.**

To summarize, here's my take on what should happen:
* For errors in the edge API, use the error responses you've defined here (for now)
* Create a JIRA for implementing error templates, specified via configuration
* Return **all** responses from the backend as-is - success, failure, etc.
* Responses from the backend should not be parsed or validated
* The RAML cannot specify schemas for any responses since they will differ depending on the 'type' query argument.

Comment (craigmcnally) [reply src/main/java/org/folio/edge/orders/OrdersHandler.java:106]: OK, after a discussion and closer code review it seems that the proposed plan above needs to be tweaked because mod-gobi currently returns text/plain.

1. If the backend returns text/plain, wrap in the XML/JSON error response format
2. Story for mod-gobi to return error XML conforming to the schema currently in use - currently returns text/plain
3. Story for edge-orders to remove the wrapping and return responses as-is (depends on the mod-gobi work above)
4. Models and scheams for gobi order responses (success) can be removed - no longer needed
5. For now, leave the examples in RAML - remove as part of #3

File: src/main/java/org/folio/edge/orders/OrdersHandler.java [hunk 8]
```diff
@@ -91,115 +88,107 @@ protected void handle(RoutingContext ctx, List<Routing> routingMapping) {
...
+        .end(respBody);
+    } else {
+      ErrorCodes errorCode = Optional.ofNullable(ErrorCodes.fromValue(status)).orElse(ErrorCodes.INTERNAL_SERVER_ERROR);
+      ResponseWrapper resp = new ResponseWrapper(new ErrorWrapper(errorCode.name(), respBody));
+      handleErrorResponse(ctx, status, resp);
+    }
+  }
+
+  private boolean isSuccessStatus(int status) {
+    return status == 200 || status == 201;
```
Comment (craigmcnally) [inline src/main/java/org/folio/edge/orders/OrdersHandler.java:173]: A better check is anything in the 2XX range... You're likely to see 204, and maybe even 202s for instance...

File: src/main/java/org/folio/edge/orders/OrdersHandler.java [hunk 9]
```diff
@@ -91,115 +88,107 @@ protected void handle(RoutingContext ctx, List<Routing> routingMapping) {
...
+  }
+
+  private void handleErrorResponse(RoutingContext ctx, int status, ResponseWrapper responseWrapper) {
+    String acceptHeader = ctx.request().getHeader(HttpHeaders.ACCEPT);
+    ctx.response().setStatusCode(status);
+    try {
+      if (APPLICATION_JSON.equals(acceptHeader)) {
```
Comment (craigmcnally) [inline src/main/java/org/folio/edge/orders/OrdersHandler.java:180]: Unfortunately I don't think this is good enough... accept can have several values and sometimes also has an encoding at the end... a straight-up equals will not work in these cases.

File: src/main/java/org/folio/edge/orders/OrdersHandler.java [hunk 10]
```diff
@@ -91,115 +88,107 @@ protected void handle(RoutingContext ctx, List<Routing> routingMapping) {
...
+        ctx.response()
+          .putHeader(HttpHeaders.CONTENT_TYPE, APPLICATION_JSON)
+          .end(responseWrapper.toJson());
+      } else {
+        ctx.response()
+          .putHeader(HttpHeaders.CONTENT_TYPE, APPLICATION_XML)
```
Comment (craigmcnally) [inline src/main/java/org/folio/edge/orders/OrdersHandler.java:186]: It might be worth adding a note to the readme and/or RAML that by default this returns application/xml...  For instance, if I pass Accept: */*, I'll get application/xml.

This is IMO the right thing to do - at least for the short term since it isn't clear to me if GOBI sends an accept header or not, and it definitely expects application/xml for error responses.

Comment (craigmcnally) [reply src/main/java/org/folio/edge/orders/OrdersHandler.java:186]: From the perspective of making this general purpose and not tied to a specific integration, we may want to eventually change the default to text/plain.  This is out of scope here though, and as mentioned above would be a breaking change that very well could break aspects of the gobi integration

File: src/test/java/org/folio/edge/orders/MainVerticleTest.java [hunk 11]
```diff
@@ -458,4 +480,39 @@ public void testPlaceOrderTimeout(TestContext context) throws MalformedApiKeyExc
...
+      .with()
+      .accept(APPLICATION_JSON)
+      .body(BODY_REQUEST_FOR_HEADER_INCONSISTENCY)
+      .post("/orders?type=GOBI&apiKey=" + apiKey)
+      .then()
+      .contentType(APPLICATION_JSON)
+      .statusCode(400)
+      .extract()
+      .response();
+  }
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/orders/MainVerticleTest.java:517]: Maybe add a test that covers the cases I mentioned earlier
* `*/*`
* multiple values
* w/ encoding

File: src/test/java/org/folio/edge/orders/MainVerticleTest.java [hunk 12]
```diff
@@ -458,4 +481,157 @@ public void testPlaceOrderTimeout(TestContext context) throws MalformedApiKeyExc
...
+      .extract()
+      .response();
+    Object error = resp.path("Error");
+
+    assertNotNull(error);
+  }
+
+  @Test
+  public void testShouldReturnTextErrorWithInternalFormatIfAcceptHeaderIsText() {
+    logger.info("=== Test place order - Success (JSON) ===");
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/orders/MainVerticleTest.java:554]: This log message is misleading

File: src/test/java/org/folio/edge/orders/MainVerticleTest.java [hunk 13]
```diff
@@ -458,4 +481,157 @@ public void testPlaceOrderTimeout(TestContext context) throws MalformedApiKeyExc
...
+    final Response resp = RestAssured
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/orders/MainVerticleTest.java:555]: if you're not doing anything with the response there's no need to extract it.

File: src/test/java/org/folio/edge/orders/MainVerticleTest.java [hunk 14]
```diff
@@ -458,4 +481,157 @@ public void testPlaceOrderTimeout(TestContext context) throws MalformedApiKeyExc
...
+      .then()
+      .contentType(TEXT_PLAIN)
+      .statusCode(400)
+      .extract()
+      .response();
+  }
+
+  @Test
+  public void testShouldReturnXMLErrorWithInternalFormatIfAcceptHeaderIsAny() {
+    logger.info("=== Test place order - Success (JSON) ===");
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/orders/MainVerticleTest.java:569]: ditto

Comment (craigmcnally) [reply src/test/java/org/folio/edge/orders/MainVerticleTest.java:569]: please fix all these log messages - there are a bunch of them.
